src/adapters/supabase/helpers/log.ts

```diff
@@ -0,0 +1,197 @@
+import { getAdapters, getBotContext, Logger } from "../../../bindings";
+import { Payload } from "../../../types";
+import { getOrgAndRepoFromPath } from "../../../utils/private";
+interface Log {
+  repo: string | null;
+  org: string | null;
+  commentId: number | undefined;
+  issueNumber: number | undefined;
+  logMessage: string;
+  errorType: levels;
+  timestamp: string;
+}
+
+export enum levels {
+  error = 0,
+  warn = 1,
+  info = 2,
+  http = 3,
+  verbose = 4,
+  debug = 5,
+  silly = 6,
+}
+
+export class GitHubLogger implements Logger {
+  private supabase = getAdapters().supabase;
+  private maxLevel;
+  private app;
+  private logEnvironment;
+  private logQueue: Log[] = []; // Your log queue
+  private maxConcurrency = 6; // Maximum concurrent requests
+  private retryDelay = 1000; // Delay between retries in milliseconds
+  private throttleCount = 0;
+  private retryLimit = 0; // Retries disabled by default
+
+  constructor(app: string, logEnvironment: string, maxLevel: keyof typeof levels, retryLimit: number) {
+    this.app = app;
+    this.logEnvironment = logEnvironment;
+    this.maxLevel = levels[maxLevel];
+    this.retryLimit = retryLimit;
+  }
+
+  async sendLogsToSupabase({ repo, org, commentId, issueNumber, logMessage, errorType, timestamp }: Log) {
+    try {
+      const { error } = await this.supabase.from("logs").insert([
+        {
+          repo_name: repo,
+          level: errorType,
+          org_name: org,
+          comment_id: commentId,
+          log_message: logMessage,
+          issue_number: issueNumber,
+          timestamp,
+        },
+      ]);
+
+      if (error) {
+        console.error("Error logging to Supabase:", error.message);
+        return;
+      }
+    } catch (error) {
+      if (error instanceof Error) {
+        console.error("An error occurred:", error.message);
+      }
+    }
+  }
+
+  async processLogs(log: Log) {
+    try {
+      await this.sendLogsToSupabase(log);
+    } catch (error) {
+      console.error("Error sending log, retrying:", error);
+      return this.retryLimit > 0 ? await this.retryLog(log) : null;
+    }
+  }
+
+  async retryLog(log: Log, retryCount = 0) {
+    if (retryCount >= this.retryLimit) {
+      console.error("Max retry limit reached for log:", log);
+      return;
+    }
+
+    await new Promise((resolve) => setTimeout(resolve, this.retryDelay));
+
+    try {
+      await this.sendLogsToSupabase(log);
+    } catch (error) {
+      console.error("Error sending log (after retry):", error);
+      await this.retryLog(log, retryCount + 1);
+    }
+  }
+
+  async processLogQueue() {
+    while (this.logQueue.length > 0) {
+      const log = this.logQueue.shift();
+      if (!log) {
+        continue;
+      }
+      await this.processLogs(log);
+    }
+  }
+
+  async throttle(fn: () => void) {
+    if (this.throttleCount >= this.maxConcurrency) {
+      return;
+    }
+
+    this.throttleCount++;
+    try {
+      await fn();
```

Comment:
Both same thing, calling fn() was just shorter while preserving `this` at the same time

Reply:
If you call `processLogQueue` directly, it still preserves `this`, isn't it?

Reply:
Yes, but then the first fn call gives way for it to be called the next time with the preserved this..

If there was no `fn` param, it can no longer be passed in the `finally` block to preserve the this

So let's leave that the way it is

Reply:
I just tried directly calling it and it works the same...I still don't understand why does this make any difference? 
`this` is just a reference to the current instance of the class and it will be available to the function either way

Reply:
> I just tried directly calling it and it works the same...I still don't understand why does this make any difference? `this` is just a reference to the current instance of the class and it will be available to the function either way

Alright, I've removed it

Reply:
> I just tried directly calling it and it works the same...I still don't understand why does this make any difference? `this` is just a reference to the current instance of the class and it will be available to the function either way

Show me how you called it directly here, the last commit i made broke the log queue

Reply:
> > I just tried directly calling it and it works the same...I still don't understand why does this make any difference? `this` is just a reference to the current instance of the class and it will be available to the function either way
> 
> Show me how you called it directly here, the last commit i made (calling it directly) broke the log queue

Reply:
```js
async throttle() {
    if (this.throttleCount >= this.maxConcurrency) {
      return;
    }

    this.throttleCount++;
    try {
      await this.processLogQueue();
    } finally {
      this.throttleCount--;
      if (this.logQueue.length > 0) {
        await this.throttle();
      }
    }
  }

  async addToQueue(log: Log) {
    this.logQueue.push(log);
    if (this.throttleCount < this.maxConcurrency) {
      await this.throttle();
    }
  }
```

Reply:
Thanks, fixed it

Reply:
> I just tried directly calling it and it works the same...I still don't understand why does this make any difference? `this` is just a reference to the current instance of the class and it will be available to the function either way

This is getting spicy